### PR TITLE
chore/refactor katalogus settings setup

### PR DESCRIPTION
**Pull request by Donnype**
## Changes
As this turns out to be a larger refactor, this PR only refactors some Katalogus settings logic and makes more use of a typed property instead of dicts everywhere. We now only call the settings endpoints that gets all settings, which is also a setup of the change I have in mind to start supporting json-schema validation for these settings.

This also fixes some other issues:
- Not handling any Katalogus request errors: we now redirect where possible and display useful info about what could be retrieved from the API and what not. E.g. when the katalogus is down and we go to a plugin detail page we now see:

image

- One less call to the katalogus to fetch the repository id when we have all plugin information available (which is the common case)

- There was some strange behavior when empty strings were added as values, as the frontend showed them as not set but the add_setting calls failed:

image

Since I think settings are truly unset when they are either `null` or not present as a key, I updated empty string support so we could see the following and the functionality is fixed:

image
 

- 404's when we try to add/edit/delete an invalid setting:

image

- 404 when trying to get a plugin that could not be found in the KATalogus:

image

As a reference, this is the current view structure now:
```mermaid
classDiagram
direction BT
    class FormView
    class OrganizationView
    class SinglePluginView
    class KATalogusView
    class PluginSettingsAddView
    class PluginEnableDisableView
    class SingleSettingView
    class PluginSettingsListView
    
    OrganizationView : organization
    OrganizationView : octopoes_api_connector
    OrganizationView : organization_member
    OrganizationView : indemnification_present
     
    SinglePluginView : katalogus_client
    SinglePluginView : plugin
    SinglePluginView : plugin_schema
    
    SingleSettingView : setting_name
    
    class PluginSettingsUpdateView
    class PluginSettingsDeleteView
    class PluginDetailView

    KATalogusView  <|--  OrganizationView
    KATalogusView  <|--  FormView
    SinglePluginView  <|--  OrganizationView
    SingleSettingView  <|--  SinglePluginView
    PluginDetailView  <|--  PluginSettingsListView
    PluginEnableDisableView  <|--  SinglePluginView
    PluginSettingsAddView  <|--  FormView
    PluginSettingsAddView  <|--  SinglePluginView
    PluginSettingsDeleteView  <|--  SingleSettingView
    PluginSettingsUpdateView  <|--  FormView
    PluginSettingsUpdateView  <|--  SingleSettingView
    PluginSettingsListView  <|--  SinglePluginView
```

This could still be improved.

This also closes  https://github.com/minvws/nl-kat-coordination/issues/151.

### CODE REVIEW

**Review by ammar92 — APPROVED** [review]
Few small remarks, but looks good in general

File: rocky/katalogus/views/plugin_enable_disable.py [hunk 1]
```diff
@@ -1,65 +1,80 @@
...
 
 @class_view_decorator(otp_required)
-class PluginEnableDisableView(OrganizationView):
-    def dispatch(self, request, *args, **kwargs):
-        self.katalogus_client = get_katalogus(self.organization.code)
-        return super().dispatch(request, *args, **kwargs)
+class PluginEnableDisableView(SinglePluginView):
+    def check_required_settings(self, settings: Dict):
+        if not self.plugin_schema:
+            return True
```
[the remark this answers is not in the document]
Comment (Donnype) [reply rocky/katalogus/views/plugin_enable_disable.py:22]: I can be `None` as well though, hence the check

File: rocky/katalogus/views/mixins.py [hunk 2]
```diff
@@ -14,20 +20,64 @@
...
+
+            raise
+        except RequestException:
+            messages.add_message(
+                self.request,
+                messages.ERROR,
+                _("Getting information for plugin {} failed. Please check the KATalogus logs.").format(plugin_id),
+            )
+
+    def dispatch(self, request, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (Donnype) [reply rocky/katalogus/views/mixins.py:53]: If we are on a plugin detail page and the Katalogus goes down or the plugin disappears, when we then submit a form with a POST we still need to handle the error this check detects. We could do a separate check on the POST handler and redirect to the GET handler on an error, and change this `dispatch` to a `get`. However, to make such checks have effect we would have to add `super()` calls in each `get()` and `post()` of child views, while child views usually don't have to override the dispatch method. And I was looking for a way for child views to not have to worry about Katalogus error handling.

[the remark this answers is not in the document]
Comment (underdarknl) [reply rocky/katalogus/views/mixins.py:53]: In general
These views should just connect to a KAT-alogus Client.  Ideally by having a More Object Oriented / orm like approach to changing PluginSettings.
```
katalogusconnection = KATalogus('https://localhost:800x')
try:
  boefje = Boefje.FromPrimary(katalogusconnection, 'NMAP')
  boefje['settings'] = {'key':'value'}
  boefje.Save()
except KatalogusLoadError:
  raise FrontendError('Cannot Load boefje NMAP') # or raise 404 in the frontend if there is no nice way to handle this in the view by providing a search, or redirecting to list with this error message on top.
except KatalogusSettingsFormatError:
  raise FrontendError('Cannot save setting changes for NMAP, your input was incorrect') # or pick apart the specific error and use that to re-render the form
except KatalogusError: 
  # this includes all non specific http errors the underlying request lib might throw, including timeouts, dns issues, cert issues, remote server errors.
  raise FrontendError('Cannot communicate with katalogus, see /health for details')
```

It should not matter to the views if they are talking to a database, a moonbase or an api. Abstract the errors into meaningful errors if you can. Or if they cannot be handled in the view, just let it be and communicate what broke generally.

[the remark this answers is not in the document]
Comment (underdarknl) [reply rocky/katalogus/views/mixins.py:53]: separate PR for the clients.

**Comment by Darwinkel** [comment]
## Checklist for functional reviewer(s):
- [x] If a non-trivial PR: This PR is properly linked to an issue on the project board;
- [x] I have checked out this branch, and successfully ran `make kat`;
- [ ] I have ran `make test-rf` and all end-to-end Robot Framework tests pass;
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended;
- [x] I confirmed that there are no unintended functional regressions in this branch;

### What works:
* Adding settings
* Editing settings
* Removing settings
* Disabling/enabling Boefjes
* No hard errors when entering various kinds of garbage data
* One-time error in the crisis room when the KATalogus is unreachable
* Empty strings != Unset

### Bug or feature?:
* The KATalogus errors in the crisis room disappear after a page refresh - accidentally switching pages causes them to be forever lost
* When attempting to open the KATalogus when the container is down, there is no fancy error handling.
image

**Comment by Donnype** [comment]
@Darwinkel @underdarknl Thanks for catching that! This shouldn't happen since that was exactly the exception I was trying to handle. Will have a look.

**Comment by Donnype** [comment]
@underdarknl @Darwinkel This turned out to be a merge conflict that should be fixed now.

@Darwinkel What exactly do you mean with this one? I believe the idea indeed is that the errors do not persist on a refresh. 

- The KATalogus errors in the crisis room disappear after a page refresh - accidentally switching pages causes them to be forever lost. Is this the desired behaviour?

**Comment by Darwinkel** [comment]
## Checklist for functional reviewer(s):
- [x] If a non-trivial PR: This PR is properly linked to an issue on the project board;
- [x] I have checked out this branch, and successfully ran `make kat`;
- [ ] I have ran `make test-rf` and all end-to-end Robot Framework tests pass;
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended;
- [x] I confirmed that there are no unintended functional regressions in this branch;

### What works:
* Disabling/enabling Boefjes
* Proper redirect, messages, and error handling when the Katalogus is down


### What doesn't work:
image

* Accessing a Boefje page when settings have been modified (settings also don't appear to take effect)

image

* Creating a new organization while the Katalogus is down gives a hard RockyError, should this be turned into a friendly error message?

**Comment by Darwinkel** [comment]
Can confirm first issue has been fixed now.

**Comment by Donnype** [comment]
@Darwinkel I will fix the second one in a different PR as this issue has been introduced in a different feature :+1:

**Comment by Donnype** [comment]
@Darwinkel https://github.com/minvws/nl-kat-coordination/pull/709

**Review by ammar92 — APPROVED** [review]

**Comment by Darwinkel** [comment]
@Donnype Rocky unit tests now fail after the most recent merge.

**Comment by Donnype** [comment]
@Darwinkel Fixed

**Review by Darwinkel — APPROVED** [review]

**Review by dekkers — APPROVED** [review]

**Pull request merged**